src/primitives/compression.cpp

```diff
@@ -0,0 +1,283 @@
+#include <primitives/compression.h>
+
+CCompressedOutPoint::CCompressedOutPoint(const uint32_t block_height, const uint32_t block_index) : m_block_height(block_height), m_block_index(block_index), m_vout(0) {};
+
+CCompressedOutPoint::CCompressedOutPoint(const uint256& txid, const uint32_t vout) : m_block_height(0), m_block_index(0), m_txid(txid), m_vout(vout) {};
+
+std::string CCompressedOutPoint::ToString() const
+{
+    return strprintf("CCompressedOutPoint(block_height=%u, block_index=%u, txid=%s, vout=%u)",
+    m_block_height,
+    m_block_index,
+    HexStr(m_txid),
+    m_vout);
+}
+
+CCompressedTxIn::CCompressedTxIn(const CTxIn& txin, const CCompressedOutPoint& out, const CScript& script_pubkey) : m_prevout(out) {
+
+    m_compressedSignature = false;
+    m_hashType = 0;
+    m_isHashStandard = false;
+
+    if ((txin.scriptSig.size() || txin.scriptWitness.stack.size()) && (out.vout() != 0xFFFFFFFF))  {
+        CScript scriptSig;
+
+        if (script_pubkey.IsPayToTaproot() && txin.scriptWitness.stack.size() == 1) {
+            m_signature = txin.scriptWitness.stack.at(0);
+            if (m_signature.size() == 65) {
+                m_hashType = m_signature[m_signature.size()-1];
+                m_isHashStandard = m_hashType == 0x00;
+                m_signature.pop_back();
+            } else m_isHashStandard = true;
+            m_compressedSignature = true;
+        } else if (script_pubkey.IsPayToPublicKeyHash()) {
+            scriptSig = txin.scriptSig;
+        } else if (txin.scriptWitness.stack.size() && script_pubkey.IsPayToWitnessKeyHash()) {
+            scriptSig = CScript(txin.scriptWitness.stack.at(0).begin(), txin.scriptWitness.stack.at(0).end());
+        } else if (script_pubkey.IsPayToScriptHash()) {
+          opcodetype opcodeRet;
+          CScriptBase::const_iterator pc = txin.scriptSig.begin();
+          std::vector<unsigned char> vec;
+          txin.scriptSig.GetOp(pc, opcodeRet, vec);
+          CScript redeemScript = CScript(vec.begin(), vec.end());
+          if (redeemScript.IsPayToWitnessKeyHash()) {
+              scriptSig = CScript(txin.scriptWitness.stack.at(0).begin(), txin.scriptWitness.stack.at(0).end());
+          }
+        }
+
+        if (scriptSig.size()) {
```

Comment:
In d86bd420e798449b2796ca49b2bd924ad83c658a:

Everywhere you call `size` to check if the vector is empty, you should call `empty` instead. e.g. `!scriptSig.empty()`.

Reply:
Fixed, Thanks!